### DISCUSSION TITLE
ObservableCollection with [ObservableProperty] and [NotifyPropertyChangedFor] does not call NotifyCanExecuteChanged() when items added/removed

### DISCUSSION THREAD

**Discussion by will-scc** [CommunityToolkit/dotnet#763] [Q&A]
I don't know if this is possible, but say I have a collection and command this like this:

```c#
[ObservableProperty]
[NotifyCanExecuteChangedFor(nameof(MyCommand))]
private ObservableCollection<MyClass> selectedItems = new();

[RelayCommand(CanExecute = nameof(MyCommandCanExecute))]
private async Task MyCommand()
{
    //Do stuff
}

private bool MyCommandCanExecute()
{
    return this.SelectedItems.Count == 1;
}
```

Adding or removing items from the collection does not trigger `NotifyCanExecuteChanged`, because `[ObservableProperty]` attribute is only for when the reference to the collection itself is changed (e.g. `SelectedItems = new();`).

This can be solved by hooking up to the `CollectionChanged` event and calling the relevant things manually:
```c#
public void SelectedItemsCollectionChangedHandler(object? value, NotifyCollectionChangedEventArgs args)
{
    if (args.Action == NotifyCollectionChangedAction.Add ||
        args.Action == NotifyCollectionChangedAction.Remove)
    {
        this.MyCommandCommand.NotifyCanExecuteChanged();
    }
}
```

Is there a better way this could be handled? Could the source generators generate this additional bit of code when [ObservableProperty] is used on a collection type?

I'd have a go myself, but I'm not sure where to start with the source generators.

**Comment by Myliak** [comment]
I am actually too interested in this. I didn't find any other solution other than yours and I don't think there is. It would be nice to support it in existing attribute or to have another Attribute to support collection changes. (e.g. [NotifyCanExecuteChangedFor] -> [NotifyCanExecutePropertyChangedFor] and [NotifyCanExecuteCollectionChangedFor])
